src/main/java/org/jboss/modules/PathUtils.java

```diff
@@ -274,6 +265,27 @@ public static String canonicalize(String path) {
         return new String(targetBuf, a + 1, length - a - 1);
     }
 
+    /**
+     * Canonicalize the given path.  Removes all {@code .} and {@code ..} segments from the path.
+     *
+     * @param path the relative or absolute possibly non-canonical path
+     * @return the canonical path
+     */
+    public static String canonicalize(String path) {
+        if (path.trim().isEmpty() || !path.contains(".")) {
+            return path;
+        }
+
+        // represents a state transitions that will be found via the state machine
+        // but if these patterns aren't contained in the string they can't be found
+        // via the state machine. (we know it already contains at least "." at this point.)
+        if (!path.contains("/.") && !path.contains("./")) {
```

Comment:
Here maybe try something like:

```java
int idx = path.indexOf('.'); // maybe reuse the result of the above comparison for efficiency
if (! (path.length() > idx + 1 && isSepChar(path.charAt(idx + 1)) || idx > 0 && isSepChar(path.charAt(idx - 1))) {
```

With:

```
private static boolean isSepChar(char ch) {
    // on non-Windows, the JIT *should* eliminate the second check as redundant
    return ch == '/' || ch == File.separatorChar;
}

Reply:
Also you'd better account for the strings `"."` and `".."`, which do need to be canonicalized but do not contain `/`.

Reply:
Ah this makes sense, I will need to add some test cases to catch things like this. I will add some of these cases and then update the code.

Reply:
I found some bugs, I'm using the code you suggested now and came up with some better test cases. I will benchmark again (takes about 20-30min).

Reply:
```
Benchmark                                                            Mode  Cnt     Score      Error   Units

# META-INF/application.properties
PathUtilsBenchmarkTest.withDotButFine                                avgt    5     0.011 ±    0.001   us/op
PathUtilsBenchmarkTest.withDotButFine:·gc.alloc.rate                 avgt    5     0.002 ±    0.012  MB/sec
PathUtilsBenchmarkTest.withDotButFine:·gc.alloc.rate.norm            avgt    5    ≈ 10⁻⁶               B/op
PathUtilsBenchmarkTest.withDotButFine:·gc.count                      avgt    5       ≈ 0             counts
PathUtilsBenchmarkTest.directWithDotButFine                          avgt    5     0.060 ±    0.001   us/op
PathUtilsBenchmarkTest.directWithDotButFine:·gc.alloc.rate           avgt    5  9168.130 ±  180.717  MB/sec
PathUtilsBenchmarkTest.directWithDotButFine:·gc.alloc.rate.norm      avgt    5    72.000 ±    0.001    B/op
PathUtilsBenchmarkTest.directWithDotButFine:·gc.count                avgt    5  1004.000             counts
PathUtilsBenchmarkTest.directWithDotButFine:·gc.time                 avgt    5   674.000                 ms
PathUtilsBenchmarkTest.originalWithDotButFine                        avgt    5     0.164 ±    0.009   us/op
PathUtilsBenchmarkTest.originalWithDotButFine:·gc.alloc.rate         avgt    5  7086.643 ±  386.548  MB/sec
PathUtilsBenchmarkTest.originalWithDotButFine:·gc.alloc.rate.norm    avgt    5   152.000 ±    0.001    B/op
PathUtilsBenchmarkTest.originalWithDotButFine:·gc.count              avgt    5   580.000             counts
PathUtilsBenchmarkTest.originalWithDotButFine:·gc.time               avgt    5  1167.000                 ms

# /this/./path/has/a/../need/to/be/canonicalized
PathUtilsBenchmarkTest.changeString                                  avgt    5     0.130 ±    0.001   us/op
PathUtilsBenchmarkTest.changeString:·gc.alloc.rate                   avgt    5  4680.515 ±   47.435  MB/sec
PathUtilsBenchmarkTest.changeString:·gc.alloc.rate.norm              avgt    5    80.000 ±    0.001    B/op
PathUtilsBenchmarkTest.changeString:·gc.count                        avgt    5   919.000             counts
PathUtilsBenchmarkTest.changeString:·gc.time                         avgt    5   496.000                 ms
PathUtilsBenchmarkTest.directChangeString                            avgt    5     0.112 ±    0.001   us/op
PathUtilsBenchmarkTest.directChangeString:·gc.alloc.rate             avgt    5  5433.140 ±   46.795  MB/sec
PathUtilsBenchmarkTest.directChangeString:·gc.alloc.rate.norm        avgt    5    80.000 ±    0.001    B/op
PathUtilsBenchmarkTest.directChangeString:·gc.count                  avgt    5   932.000             counts
PathUtilsBenchmarkTest.directChangeString:·gc.time                   avgt    5   514.000                 ms
PathUtilsBenchmarkTest.originalChangeString                          avgt    5     0.173 ±    0.024   us/op
PathUtilsBenchmarkTest.originalChangeString:·gc.alloc.rate           avgt    5  8485.424 ± 1181.296  MB/sec
PathUtilsBenchmarkTest.originalChangeString:·gc.alloc.rate.norm      avgt    5   192.000 ±    0.001    B/op
PathUtilsBenchmarkTest.originalChangeString:·gc.count                avgt    5   764.000             counts
PathUtilsBenchmarkTest.originalChangeString:·gc.time                 avgt    5   870.000                 ms

# /this/path/has/no/need/to/be/canonicalized
PathUtilsBenchmarkTest.noChangeString                                avgt    5     0.012 ±    0.001   us/op
PathUtilsBenchmarkTest.noChangeString:·gc.alloc.rate                 avgt    5     0.002 ±    0.012  MB/sec
PathUtilsBenchmarkTest.noChangeString:·gc.alloc.rate.norm            avgt    5    ≈ 10⁻⁶               B/op
PathUtilsBenchmarkTest.noChangeString:·gc.count                      avgt    5       ≈ 0             counts
PathUtilsBenchmarkTest.directNoChangeString                          avgt    5     0.094 ±    0.003   us/op
PathUtilsBenchmarkTest.directNoChangeString:·gc.alloc.rate           avgt    5  7141.097 ±  249.419  MB/sec
PathUtilsBenchmarkTest.directNoChangeString:·gc.alloc.rate.norm      avgt    5    88.000 ±    0.001    B/op
PathUtilsBenchmarkTest.directNoChangeString:·gc.count                avgt    5  1009.000             counts
PathUtilsBenchmarkTest.directNoChangeString:·gc.time                 avgt    5   617.000                 ms
PathUtilsBenchmarkTest.originalNoChangeString                        avgt    5     0.189 ±    0.051   us/op
PathUtilsBenchmarkTest.originalNoChangeString:·gc.alloc.rate         avgt    5  7794.562 ± 2195.670  MB/sec
PathUtilsBenchmarkTest.originalNoChangeString:·gc.alloc.rate.norm    avgt    5   192.000 ±    0.001    B/op
PathUtilsBenchmarkTest.originalNoChangeString:·gc.count              avgt    5   713.000             counts
PathUtilsBenchmarkTest.originalNoChangeString:·gc.time               avgt    5  1104.000                 ms

# /../../../../../.././.thing/../../././../
PathUtilsBenchmarkTest.manyChangeString                              avgt    5     0.124 ±    0.003   us/op
PathUtilsBenchmarkTest.manyChangeString:·gc.alloc.rate               avgt    5  2959.892 ±   82.816  MB/sec
PathUtilsBenchmarkTest.manyChangeString:·gc.alloc.rate.norm          avgt    5    48.000 ±    0.001    B/op
PathUtilsBenchmarkTest.manyChangeString:·gc.count                    avgt    5  1058.000             counts
PathUtilsBenchmarkTest.manyChangeString:·gc.time                     avgt    5   337.000                 ms
PathUtilsBenchmarkTest.directManyChangeString                        avgt    5     0.070 ±    0.001   us/op
PathUtilsBenchmarkTest.directManyChangeString:·gc.alloc.rate         avgt    5  5207.492 ±  103.130  MB/sec
PathUtilsBenchmarkTest.directManyChangeString:·gc.alloc.rate.norm    avgt    5    48.000 ±    0.001    B/op
PathUtilsBenchmarkTest.directManyChangeString:·gc.count              avgt    5   885.000             counts
PathUtilsBenchmarkTest.directManyChangeString:·gc.time               avgt    5   573.000                 ms
PathUtilsBenchmarkTest.originalManyChangeString                      avgt    5     0.157 ±    0.009   us/op
PathUtilsBenchmarkTest.originalManyChangeString:·gc.alloc.rate       avgt    5  7384.494 ±  418.206  MB/sec
PathUtilsBenchmarkTest.originalManyChangeString:·gc.alloc.rate.norm  avgt    5   152.000 ±    0.001    B/op
PathUtilsBenchmarkTest.originalManyChangeString:·gc.count            avgt    5   610.000             counts
PathUtilsBenchmarkTest.originalManyChangeString:·gc.time             avgt    5  1087.000                 ms

# META-INF/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/.hidden
PathUtilsBenchmarkTest.longNoChangeCase                              avgt    5     1.783 ±    0.030   us/op
PathUtilsBenchmarkTest.longNoChangeCase:·gc.alloc.rate               avgt    5  1882.206 ±   32.076  MB/sec
PathUtilsBenchmarkTest.longNoChangeCase:·gc.alloc.rate.norm          avgt    5   440.000 ±    0.002    B/op
PathUtilsBenchmarkTest.longNoChangeCase:·gc.count                    avgt    5   700.000             counts
PathUtilsBenchmarkTest.longNoChangeCase:·gc.time                     avgt    5   206.000                 ms
PathUtilsBenchmarkTest.directLongNoChangeCase                        avgt    5     1.705 ±    0.016   us/op
PathUtilsBenchmarkTest.directLongNoChangeCase:·gc.alloc.rate         avgt    5  1968.186 ±   18.589  MB/sec
PathUtilsBenchmarkTest.directLongNoChangeCase:·gc.alloc.rate.norm    avgt    5   440.000 ±    0.003    B/op
PathUtilsBenchmarkTest.directLongNoChangeCase:·gc.count              avgt    5   723.000             counts
PathUtilsBenchmarkTest.directLongNoChangeCase:·gc.time               avgt    5   207.000                 ms
PathUtilsBenchmarkTest.originalLongNoChangeCase                      avgt    5     1.849 ±    0.030   us/op
PathUtilsBenchmarkTest.originalLongNoChangeCase:·gc.alloc.rate       avgt    5  5150.149 ±   83.746  MB/sec
PathUtilsBenchmarkTest.originalLongNoChangeCase:·gc.alloc.rate.norm  avgt    5  1248.000 ±    0.003    B/op
PathUtilsBenchmarkTest.originalLongNoChangeCase:·gc.count            avgt    5   632.000             counts
PathUtilsBenchmarkTest.originalLongNoChangeCase:·gc.time             avgt    5   819.000                 ms

# META-INF/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/a/.
PathUtilsBenchmarkTest.longWorstCase                                 avgt    5     1.732 ±    0.015   us/op
PathUtilsBenchmarkTest.longWorstCase:·gc.alloc.rate                  avgt    5  1903.240 ±   16.433  MB/sec
PathUtilsBenchmarkTest.longWorstCase:·gc.alloc.rate.norm             avgt    5   432.000 ±    0.001    B/op
PathUtilsBenchmarkTest.longWorstCase:·gc.count                       avgt    5   749.000             counts
PathUtilsBenchmarkTest.longWorstCase:·gc.time                        avgt    5   211.000                 ms
PathUtilsBenchmarkTest.directLongWorstCase                           avgt    5     1.700 ±    0.014   us/op
PathUtilsBenchmarkTest.directLongWorstCase:·gc.alloc.rate            avgt    5  1938.810 ±   16.285  MB/sec
PathUtilsBenchmarkTest.directLongWorstCase:·gc.alloc.rate.norm       avgt    5   432.000 ±    0.003    B/op
PathUtilsBenchmarkTest.directLongWorstCase:·gc.count                 avgt    5   729.000             counts
PathUtilsBenchmarkTest.directLongWorstCase:·gc.time                  avgt    5   211.000                 ms
PathUtilsBenchmarkTest.originalLongWorstCase                         avgt    5     1.540 ±    0.014   us/op
PathUtilsBenchmarkTest.originalLongWorstCase:·gc.alloc.rate          avgt    5  6102.714 ±   55.424  MB/sec
PathUtilsBenchmarkTest.originalLongWorstCase:·gc.alloc.rate.norm     avgt    5  1232.000 ±    0.002    B/op
PathUtilsBenchmarkTest.originalLongWorstCase:·gc.count               avgt    5   772.000             counts
PathUtilsBenchmarkTest.originalLongWorstCase:·gc.time                avgt    5   978.000                 ms
```

Broad strokes: there aren't any surprises here. We pay a cost for skipping out early and we pay a price for the ThreadLocal. I think this produces tighter performance in most cases, fewer allocations in most cases, and only seems to cost about 10%-14% in the worst case. In cases that clearly don't need modification this is much faster.